test/balancerd/mzcompose.py

```diff
@@ -186,6 +190,126 @@ def workflow_http(c: Composition) -> None:
+    def create_proxy_protocol_v2_header(
+        client_ip: str, client_port: int, server_ip: str, server_port: int
+    ):
+        # Signature for Proxy Protocol v2
+        signature = b"\r\n\r\n\x00\r\nQUIT\n"
+        # Version and command (0x21 means version 2, PROXY command)
+        version_and_command = 0x21
+        # Address family and protocol (0x11 means INET (IPv4) + STREAM (TCP))
+        family_and_protocol = 0x11
+        # Source and destination address are sent as bytes.
+        src_addr = socket.inet_aton(client_ip)
+        dst_addr = socket.inet_aton(server_ip)
+        # Pack ports into 2-byte unsigned integers
+        src_port = struct.pack("!H", client_port)
+        dst_port = struct.pack("!H", server_port)
+        # Length of the address information (IPv4(4*2) + ports(1*2) = 12 bytes)
+        addr_len = struct.pack("!H", 12)
+        # Construct the final Proxy Protocol v2 header
+        header = (
+            signature
+            + struct.pack("!BB", version_and_command, family_and_protocol)
+            + addr_len
+            + src_addr
+            + dst_addr
+            + src_port
+            + dst_port
+        )
+
+        return header
+
+    with contextlib.closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as sock:
+        sock.connect(("127.0.0.1", materialize_port))
+
+        # Pick an ip we couldn't normal connect from and trick envd into
+        # thinking we're connecting with
+        proxy_header = create_proxy_protocol_v2_header(
+            "1.1.1.1", 1111, "127.0.0.1", 1111
+        )
+        # Make an http request over the socket
+
+        json_data = {
+            "query": "select client_address from mz_internal.mz_sessions where connection_id = pg_backend_pid();"
+        }
+        json_data = json.dumps(json_data)
+        content_length = len(json_data.encode("utf-8"))
+        http_sql_query_request = dedent(
+            f"""\
+            POST /api/sql HTTP/1.1\r
+            Host: 127.0.0.1:{materialize_port}\r
+            Authorization: Basic {OTHER_USER}:{app_password(OTHER_USER)}\r
+            Content-Type: application/json\r
+            Content-Length: {content_length}\r
+            \r
+            {json_data}"""
+        )
+        # sock.sendall(h+'POST /api/sql HTTP/1.1\r\nContent-type: application/json\r\nContent-Length: 39\r\nAuthorization: Basic materialize:\r\n\r\n{"query": "SELECT * from mz_clusters;"}'.encode('utf-8'))
+        sock.sendall(proxy_header + http_sql_query_request.encode("utf-8"))
+
+        # read and parse the response
+        body_separator = "\r\n\r\n"
+        tcp_resp = sock.recv(8192)
+        resp_split = tcp_resp.split(body_separator.encode("utf-8"))
+        assert (
+            len(resp_split) > 1
+        ), f"expected response with header and body, found: {resp_split}"
+        body = resp_split[1]
+        # assert that we tricked environmentd
+        assert json.loads(body)["results"][0]["rows"][0][0] == "1.1.1.1"
+
```

Comment:
Can you also add a test that ensures that the balancer rejects SQL connections that attempt to set the `mz_forwarded_for` option?

Reply:
Do you know of a good way to test this? I don't see how I can input these params anywhere in pg8000, and I don't see any mention of `mz_connection_uuid` (the other param the balancer uses) anywhere in the test.